server/app/models/stack.rb

```diff
@@ -11,6 +11,8 @@ class Stack
   has_many :stack_revisions, dependent: :destroy
   has_many :stack_deploys, dependent: :destroy
   has_many :grid_services
+  has_many :volumes
+  embeds_many :external_volumes
```

Comment:
If that feels too complicated, and it probably is.... just make all of the volume references use a layer of indirection? Keep the `Grid`, `Stack` and `Volume` references as they are, but drop the `ExternalVolumes` completely, and then refactor the `GridService` to parse out the volumes as embedded `ServiceVolume` models:

```ruby
class GridService
  embeds_many :volumes, class: ServiceVolume
end

class ServiceVolume
  embedded_in :grid_service

  field :bind_mount, type: String # optional host bind mount
  field :path, type: String # mount path in the container
  field :flags, type: String # optional flags

  belongs_to :volume # optional
end
```

with associated `has_many :service_volumes` on the `Volume` model.